tls/s2n_cipher_suites.c

```diff
@@ -148,6 +149,7 @@ struct s2n_cipher_suite s2n_rsa_with_rc4_128_md5 = /* 0x00,0x04 */ {
     .name = "RC4-MD5",
     .iana_value = { TLS_RSA_WITH_RC4_128_MD5 },
     .key_exchange_alg = &s2n_rsa,
+    .signature_alg = S2N_SIGNATURE_RSA,
```

Comment:
Is this a bit counter-intuitive? The TLS_RSA_WITH_* cipher suites don't actually use RSA signatures, they use RSA encryption for key exchange, and then HMAC. The DHE/EDH ones do use RSA signatures though. Would these cipher suites break if this were left at null or NONE or something?

Reply:
The relationship between cipher suites, certificates, and signature algorithms was the most difficult part for me to wrap my head around. I think in TLS 1.1, there was a requirement that the algorithm used to sign a certificate must be the same as the key type. I had interpreted that to mean that TLS_RSA_WITH_* cipher suites required certificates that had an RSA signature and that is sort of where this came from.

I tried to rethink what this field was modeling and where it was needed in the handshake. It is used in two places: to check that a received cert is compatible with the cipher suite, and to set a default signature/hash algorithm pair if ECDSA signatures will be used (also related to the cipher suite). I thought of renaming this to represent an authentication method rather than a signature algorithm. For some cipher suites, a signature would be used for authentication, but for the TLS_RSA_WITH_* the authentication would be implicit in the RSA key exchange. I'm very open to thoughts on this approach.